fendermint/vm/interpreter/src/chain.rs

```diff
@@ -177,32 +217,87 @@ where
                     }
 
                     // We can use the same result type for now, it's isomorphic.
-                    Ok(((pool, state), ChainMessageApplyRet::Signed(ret)))
+                    Ok(((pool, provider, state), ChainMessageApplyRet::Signed(ret)))
                 }
                 IpcMessage::BottomUpExec(_) => {
                     todo!("#197: implement BottomUp checkpoint execution")
                 }
-                IpcMessage::TopDown => {
-                    todo!("implement TopDown handling; this is just a placeholder")
+                IpcMessage::TopDownExec(p) => {
+                    if !provider.is_enabled() {
+                        return Err(anyhow!(
+                            "cannot execute IPC top-down message: parent provider disabled"
+                        ));
+                    }
+
+                    // commit parent finality first
+                    let finality = IPCParentFinality {
+                        height: p.height as u64,
+                        block_hash: p.block_hash,
+                    };
+                    let msg = encode_to_fvm_implicit(&encode_commit_parent_finality_call(
+                        finality.clone(),
+                    )?)?;
+                    let (state, ret) = self
+                        .inner
+                        .deliver(state, VerifiableMessage::NotVerify(msg))
+                        .await?;
+                    if ret.is_err() {
+                        // TODO: how to handle error here? Do we return Err or stash them into one ret?
+                        todo!()
+                    }
+
+                    // stash validator changes
+
+                    // error happens if we cannot get the validator set from ipc agent after retries
+                    let validator_changes = provider.validator_changes(p.height as u64).await?;
+                    let msg = encode_to_fvm_implicit(&encode_store_validator_changes_call(
+                        validator_changes,
+                    )?)?;
+                    let (state, ret) = self
+                        .inner
+                        .deliver(state, VerifiableMessage::NotVerify(msg))
+                        .await?;
+                    if ret.is_err() {
+                        // TODO: how to handle error here? Do we return Err or stash them into one ret?
```

Comment:
What can be a legitimate legitimate reasons for this to fail?

So again we could crash saying this must be a bug, since these changes have been added to the parent ledger already, how can we _not_ add them on the child? I'm not sure how we could fix the contract though, if there was a bug, but maybe the bug is in the syncer?

Or we can choose to take it on the chin and `return Ok(ret)`, however the problem is that we have already committed the finality itself, so the syncer will think it doesn't have to re-fetch these changes, and then we have two possibilities:
* We have a bug in the syncer of this node instance, and everyone else applied the changes, in which case it's a consensus failure
* Nobody applied these changes, in which case it creates a gap in the ledger that might not exist on the parent, which can lead to unaccepted checkpoints

Quite possibly the best thing would be to apply all 3 changes as one atomic call to the contracts. I don't know about a way to selectively rolling back first transaction at this point.

Reply:
I would choose to crash for the same reason I mentioned [here](https://github.com/consensus-shipyard/fendermint/pull/310/files#r1352782553).